### PR TITLE
Feature/cmp 1199 s3 bucket browser fix

### PR DESCRIPTION
**Pull request by tropotorres**
Engineer - @tropotorres 
Ticket - [CMP-1199 Residual fixes for S3-Bucket-Browser applet implementation on HUI and CUI](https://cloudbolt.atlassian.net/browse/CMP-1199)

TASK LIST
- [x] Update s3-bucket-browser with IWHs
- [x] Include IWH code for review
- [x] Update with new xui-packager 1.1.2
In file `xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py`
- [x] Handling discovered S3_bucket resources (line 264 in above file)
- [x] Handling boto3 amazon errors from undefined region (line 255 in same file)
- [x] Handling UI feedback

### CODE REVIEW

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: should this really be an error, or just a .warning() or .info()?

**Review by bernardthered — COMMENTED** [review]
Just reviewed a little part of this, still figuring out how one would review it in earnest. Will discuss in Slack.

File: xui/additional_files_unzipped/s3_upload_object/s3_upload_object/cb_plugin_214_HXw030a_KRV1EVv.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_upload_object/s3_upload_object/cb_plugin_214_HXw030a_KRV1EVv.py]: one thing I recommend anytime our code blanket catches all exceptions: 
logger.exception("what it was attempting when it failed")

That will print the full TB (to the application.log file on prod CBs), which could be extremely useful in our t-shooting. I think the full TB is lost in the current exception handling.

https://docs.python.org/3/library/logging.html#logging.Logger.exception

File: xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py]: Doing these two lines in one line with logger.exception() would be better. https://docs.python.org/3/library/logging.html#logging.Logger.exception

It looks like you're passing the generic Exception class to logger here, which will log something like this:

```
>>> print(Exception)
<class 'Exception'>
```

File: xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py]: Good idea to explain **why** in the comment!

It would be nicer if we could call S3 to check the status so we could wait <5s, but that may not be possible.

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py:206]: All of these kinds of print & log statements would be better as logger.exception()

https://docs.python.org/3/library/logging.html#logging.Logger.exception

File: xui/additional_files_unzipped/s3_upload_folder/s3_upload_folder/cb_plugin_215_r3moU3W_ju0c8Nn.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_upload_folder/s3_upload_folder/cb_plugin_215_r3moU3W_ju0c8Nn.py]: This print statement should be removed or replaced w/ a logger.info() or something like that.

File: xui/additional_files_unzipped/s3_upload_object/s3_upload_object/cb_plugin_214_HXw030a_KRV1EVv.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_upload_object/s3_upload_object/cb_plugin_214_HXw030a_KRV1EVv.py]: Ideally there should be no print() statements in production code.

File: xui/additional_files_unzipped/s3_download_item/s3_download_item/cb_plugin_220_75iqBCr_fBo9rx3.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_download_item/s3_download_item/cb_plugin_220_75iqBCr_fBo9rx3.py]: some of these files seem to be missing newlines at the end of file. I would recommend running `black` on all the .py files.

**Review by bernardthered — CHANGES_REQUESTED** [review]
Looks good overall, the exception handling code should be improved and a few other code cleanup items needed before PR approval.```

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: i think .warning() since its something that you can fix

File: xui/additional_files_unzipped/s3_upload_folder/s3_upload_folder/cb_plugin_215_r3moU3W_ju0c8Nn.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_upload_folder/s3_upload_folder/cb_plugin_215_r3moU3W_ju0c8Nn.py]: I feel like that was leftover code troubleshooting, I'm going to remove for now

File: xui/additional_files_unzipped/s3_upload_object/s3_upload_object/cb_plugin_214_HXw030a_KRV1EVv.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_upload_object/s3_upload_object/cb_plugin_214_HXw030a_KRV1EVv.py]: must be leftover, removed!

File: xui/additional_files_unzipped/s3_download_item/s3_download_item/cb_plugin_220_75iqBCr_fBo9rx3.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_download_item/s3_download_item/cb_plugin_220_75iqBCr_fBo9rx3.py]: I ran the checks and formatters before I unzipped things, is possibly why
Will get it fixed

File: xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py]: I did not write this code but I'll check if that seems possible

Comment (bernardthered) [reply xui/additional_files_unzipped/s3_promote_version/s3_promote_version/cb_plugin_222_4jAjJ9T_VOdQK9k.py]: Not worth the extra work at this point, it would introduce a bunch of complexity to the code and better to keep it simple for now.

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: It would be good to mention that this is for fetching versioning info, so they know what to expect to work and what not to work if their buckets are not in the default region.

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: will this raise another exception if the bucket isn't in the default region? if so, the code should probably catch that and return False (assuming the rest of the functionality works when we can't figure out the region)

File: xui/additional_files_unzipped/s3_get_versions/s3_get_versions/cb_plugin_218_OrDfUdw_vHnOPmv_SneuueF.py
Comment (bernardthered) [inline xui/additional_files_unzipped/s3_get_versions/s3_get_versions/cb_plugin_218_OrDfUdw_vHnOPmv_SneuueF.py]: All these cases of error handling you're doing in these IWHs are a great pattern of showing a basic error msg to the end user and printing the TB to the logs. All this error handling will pay off when customers hit errors (which they will, very frequently).

**Review by bernardthered — CHANGES_REQUESTED** [review]
This is getting very close! Just a couple last small requests

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: So learned something interesting testing this with random regions.  Changing the `default_region` to `us-west-1` didn't error, and correctly found the versioning status for the bucket I was looking at (`ui-demo-experience` in `us-east-1` 

Not only that by using that same connection I could look up info like `demo-shantam` which is in `ap-south-1`

(I should add that my RH for AWS only has `us-east-1` and `us-west-2` included
...which means you don't have to initialize the connection with the specific region that the resource is in right now, just a region that is supported by the AWS connection
Screenshot 2023-12-18 at 2 10 51 PM
So using that theory i set the default_region to `ap-east-1` and then got this 500 error: 

```
botocore.exceptions.ClientError: An error occurred (IllegalLocationConstraintException) when calling the GetBucketVersioning operation: The ap-south-1 location constraint is incompatible for the region specific endpoint this request was sent to.
```
So I think we don't need another error message for now because most errors after it tries a valid default_region are going to be botocore errors, I think

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: I have this in there already 
> "As a backup, attempt to find versioning info using the default region" 

but I will try and make this clearer and more accurate

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (bernardthered) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: Nice exploration! So does using us-east-1 work for all buckets everywhere? If so, can we just remove all the logic about trying to read custom fields and figure out which region to use, and just always use us-east-1?

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: Screenshot 2023-12-18 at 2 26 57 PM

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: Not if their account isn't enabled for 'us-east-1', which presumably is a possibility. It still makes sense to me to try and use what should be a valid user-specified input to attempt this connection

Comment (bernardthered) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: ok, cool. I'm good with it.

File: xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py
Comment (bernardthered) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: You have that in a comment, but users cannot see that when they are looking at the log file. They may be able to trace the issue back to the IWH in question and find that comment, but most will not succeed in that. Any info that would be helpful to a CB Admin in t-shooting should be moved from comment --> log message.

Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: WHOOPS sorry I also added this at the start of the function
Screenshot 2023-12-18 at 2 35 10 PM

Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: ah I can make those log messages as well

Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: OK new version
Screenshot 2023-12-18 at 2 38 28 PM

Comment (tropotorres) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: Screenshot 2023-12-18 at 2 51 33 PM


ok log when there's no `s3_bucket-region` but default works
```
2023-12-18 19:49:48,712 [INFO] shared_modules.aws_bucket_browser_operations: Checking if Bucket Versioning is enabled for the bucket 'ui-demo-experience' 
2023-12-18 19:49:48,718 [WARNING] shared_modules.aws_bucket_browser_operations: CustomField 's3_bucket_region' does not exist
2023-12-18 19:49:48,719 [INFO] shared_modules.aws_bucket_browser_operations: This error can arise when s3 buckets are added from discovery and their s3_bucket_region parmeter is not set.
2023-12-18 19:49:48,719 [INFO] shared_modules.aws_bucket_browser_operations: As a backup, attempting to connect to the AWS Resource client using the default region='us-east-1'.
2023-12-18 19:49:50,127 [INFO] shared_modules.aws_bucket_browser_operations: Connected to the AWS Resource client
```

And new log when the default also does not work (aka botocore error)
```
2023-12-18 19:47:49,178 [INFO] shared_modules.aws_bucket_browser_operations: Checking if Bucket Versioning is enabled for the bucket 'ui-demo-experience' 
2023-12-18 19:47:49,186 [WARNING] shared_modules.aws_bucket_browser_operations: CustomField 's3_bucket_region' does not exist
2023-12-18 19:47:49,187 [INFO] shared_modules.aws_bucket_browser_operations: This error can arise when s3 buckets are added from discovery and their s3_bucket_region parmeter is not set.
2023-12-18 19:47:49,191 [INFO] shared_modules.aws_bucket_browser_operations: As a backup, attempting to connect to the AWS Resource client using the default region='ap-east-1'.
2023-12-18 19:47:51,052 [INFO] common.methods: Exception in plug-in S3 Browser, line 111:
  "versioning_enabled": operations.is_versioning_enabled(s3_client,resource_id,default_region=bucket_location),
ClientError: An error occurred (IllegalLocationConstraintException) when calling the GetBucketVersioning operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.
2023-12-18 19:47:51,054 [ERROR] api.v3.errors: Unhandled exception in the CB API code:
Traceback (most recent call last):
  File "/var/www/html/cloudbolt/static/uploads/shared_modules/aws_bucket_browser_operations.py", line 254, in is_versioning_enabled
    region_attribute = resource.attributes.get(field_id=s3rcf.id)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/ec2-user/Git/cloudbolt/src/orders/models.py", line 245, in get
    return super(CustomFieldValueQuerySet, self).get(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 435, in get
    raise self.model.DoesNotExist(
orders.models.CustomFieldValue.DoesNotExist: CustomFieldValue matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 502, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/ec2-user/Git/cloudbolt/src/cbhooks/api/v3/viewsets/inbound_web_hook.py", line 93, in run
    ret = iwh.run_hook(method=request.method, parameters=parameters, files=files)
  File "/home/ec2-user/Git/cloudbolt/src/cbhooks/models.py", line 2521, in run_hook
    return self.hook.cast().run_without_retries(
  File "/home/ec2-user/Git/cloudbolt/src/cbhooks/models.py", line 3683, in run_without_retries
    return_value = callable_func(*args, **kwargs)
  File "<string>", line 111, in inbound_web_hook_post
  File "/var/www/html/cloudbolt/static/uploads/shared_modules/aws_bucket_browser_operations.py", line 272, in is_versioning_enabled
    if s3_bucket.BucketVersioning(resource.name).status == "Enabled":
  File "/usr/local/lib/python3.9/site-packages/boto3/resources/factory.py", line 339, in property_loader
    self.load()
  File "/usr/local/lib/python3.9/site-packages/boto3/resources/factory.py", line 505, in do_action
    response = action(self, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
  File "/usr/local/lib/python3.9/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.9/site-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (IllegalLocationConstraintException) when calling the GetBucketVersioning operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.
```

Comment (bernardthered) [reply xui/additional_files_unzipped/s3_browser/s3_browser/aws_bucket_browser_operations/aws_bucket_browser_operations.py]: The success & error cases look good there!

**Review by bernardthered — APPROVED** [review]
Looks good, let's ship it!

:shipit:

**Pull request merged**
